### PR TITLE
Add Application Open and Application Backgrounded Events

### PR DESCRIPTION
**Pull request by carlos-kelly**
Addresses LIB-1156 and LIB-1143.

### CODE REVIEW

File: analytics/src/main/java/com/segment/analytics/Analytics.java [hunk 1]
```diff
@@ -318,11 +322,27 @@ public void onActivityStarted(Activity activity) {
...
+                    .putValue("build", currentBuild));
           }
 
           @Override
           public void onActivityPaused(Activity activity) {
             runOnMainThread(IntegrationOperation.onActivityPaused(activity));
+            if (shutdown) {
+              return;
+            }
+            track(
```
Comment (f2prateek) [inline analytics/src/main/java/com/segment/analytics/Analytics.java]: Same as https://github.com/segmentio/analytics-ios/pull/831#discussion_r304657458 - I don't believe any properties should be collected here, as per https://segment.com/docs/spec/mobile/#application-backgrounded.

**Comment by carlos-kelly** [comment]
That is a good point. Since we don't have access to `android.arch.lifecycle` I could use the atomic boolean like we did for `onActivityCreated` to ensure it only fires for the first activity. Otherwise the user would have to manually add the tracking.

**Comment by f2prateek** [comment]
👍 I think there are tradeoffs of both approaches:

* `android.arch.lifecycle` approach: likely more reliable as they've handled edge cases for us, but requires adding a new dependency.
* boolean approach: likely less reliable as we might not account properly for all edge cases, but doesn't require a dependency.

I think the boolean approach is best for now as it mimics what we already do, but we should consider the `android.arch.lifecycle` approach in the future to make this more bullet proof.

**Comment by carlos-kelly** [comment]
Sounds good. I will update that!

**Comment by carlos-kelly** [comment]
It looks like it is part of Jetpack AndroidX so there are other dependencies to add when using `android.arch.lifecycle`. https://developer.android.com/jetpack/androidx/releases/lifecycle

**Comment by f2prateek** [comment]
Oof that sucks - yeah I think the boolean approach is definitely the right one for now!

File: analytics/src/main/java/com/segment/analytics/Analytics.java [hunk 2]
```diff
@@ -284,14 +285,25 @@ public void run() {
...
+    PackageInfo packageInfo = getPackageInfo(application);
+    final String currentVersion = packageInfo.versionName;
+    final int currentBuild = packageInfo.versionCode;
+
     activityLifecycleCallback =
         new Application.ActivityLifecycleCallbacks() {
           final AtomicBoolean trackedApplicationLifecycleEvents = new AtomicBoolean(false);
+          final AtomicInteger numberOfActivities = new AtomicInteger(1);
+          final AtomicBoolean isChangingActivityConfigurations = new AtomicBoolean(false);
+          final AtomicBoolean firstLaunch = new AtomicBoolean(false);
```
Comment (carlos-kelly) [inline analytics/src/main/java/com/segment/analytics/Analytics.java]: Android has new application lifecycle callbacks via `androidx.lifecycle`, but we can't use them due to the dependency requirements AndroidX places. Additionally that breaks the React Native wrapper as AndroidX deps won't work with new RN projects out of the box. We use the traditional method of activity reference counting to detect if we are launching a new activity or resuming to foreground.

File: analytics/src/main/java/com/segment/analytics/Analytics.java [hunk 3]
```diff
@@ -284,14 +285,25 @@ public void run() {
...
 
           @Override
           public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
+            runOnMainThread(IntegrationOperation.onActivityCreated(activity, savedInstanceState));
+
             if (!trackedApplicationLifecycleEvents.getAndSet(true)
                 && shouldTrackApplicationLifecycleEvents) {
+              numberOfActivities.set(0);
+              firstLaunch.set(true);
```
Comment (carlos-kelly) [inline analytics/src/main/java/com/segment/analytics/Analytics.java]: These reset the atomic variables for native Android apps. `onActivityCreated` is not called when this library is used by the React Native wrapper.

File: analytics/src/test/java/com/segment/analytics/AnalyticsTest.java [hunk 4]
```diff
@@ -896,19 +883,6 @@ protected boolean matchesSafely(TrackPayload payload) {
                         payload.properties().getInt("build", -1) == 101;
                   }
                 }));
-    verify(integration)
```
Comment (f2prateek) [inline analytics/src/test/java/com/segment/analytics/AnalyticsTest.java]: why was this removed?

File: analytics/src/test/java/com/segment/analytics/AnalyticsTest.java [hunk 5]
```diff
@@ -796,19 +796,6 @@ protected boolean matchesSafely(TrackPayload payload) {
                         payload.properties().getInt("build", -1) == 100;
                   }
                 }));
-    verify(integration)
```
Comment (f2prateek) [inline analytics/src/test/java/com/segment/analytics/AnalyticsTest.java]: why was this removed?

File: analytics/src/main/java/com/segment/analytics/Analytics.java [hunk 6]
```diff
@@ -318,6 +329,19 @@ public void onActivityStarted(Activity activity) {
...
+
+            if (shouldTrackApplicationLifecycleEvents
+                && numberOfActivities.incrementAndGet() == 1
+                && !isChangingActivityConfigurations.get()) {
+
+              Properties properties = new Properties();
+              if (firstLaunch.get()) {
+                properties.putValue("version", currentVersion).putValue("build", currentBuild);
+              }
+              properties.putValue("from_background", !firstLaunch.getAndSet(false));
```
Comment (carlos-kelly) [inline analytics/src/main/java/com/segment/analytics/Analytics.java]: On first launch we want to ensure `from_background` is false, subsequent calls, this will be true.

File: analytics/src/test/java/com/segment/analytics/AnalyticsTest.java (hunk 5 above)
Comment (carlos-kelly) [reply analytics/src/test/java/com/segment/analytics/AnalyticsTest.java]: Both were removed as `ApplicationOpened` is no longer called during `trackApplicationLifecycleEventsInstalled`. It is now called in `onActivityResumed` and depending if it's a cold start or not, the correct Properties are passed.

File: analytics/src/test/java/com/segment/analytics/AnalyticsTest.java [hunk 7]
```diff
@@ -1050,6 +1023,205 @@ protected boolean matchesSafely(Application.ActivityLifecycleCallbacks item) {
...
+    verify(integration)
+        .track(
+            argThat(
+                new NoDescriptionMatcher<TrackPayload>() {
+                  @Override
+                  protected boolean matchesSafely(TrackPayload payload) {
+                    return payload.event().equals("Application Opened")
+                        && payload.properties().getString("version").equals("1.0.0")
+                        && payload.properties().getInt("build", -1) == 100
+                        && !payload.properties().getBoolean("from_background", true);
```
Comment (carlos-kelly) [inline analytics/src/test/java/com/segment/analytics/AnalyticsTest.java]: This is the first integration test that ensures that on first start `from_background` is false, the correct build params are part of the cold start `Application Opened` tracking event.

File: analytics/src/test/java/com/segment/analytics/AnalyticsTest.java [hunk 8]
```diff
@@ -1050,6 +1023,205 @@ protected boolean matchesSafely(Application.ActivityLifecycleCallbacks item) {
...
+    callback.get().onActivityResumed(null);
+    callback.get().onActivityStopped(backgroundedActivity);
+
+    verify(integration)
+        .track(
+            argThat(
+                new NoDescriptionMatcher<TrackPayload>() {
+                  @Override
+                  protected boolean matchesSafely(TrackPayload payload) {
+                    return payload.event().equals("Application Backgrounded");
```
Comment (carlos-kelly) [inline analytics/src/test/java/com/segment/analytics/AnalyticsTest.java]: This is the second integration test that launches the app then sends it to background.

File: analytics/src/test/java/com/segment/analytics/AnalyticsTest.java [hunk 9]
```diff
@@ -1050,6 +1023,205 @@ protected boolean matchesSafely(Application.ActivityLifecycleCallbacks item) {
...
+                }));
+
+    verify(integration)
+        .track(
+            argThat(
+                new NoDescriptionMatcher<TrackPayload>() {
+                  @Override
+                  protected boolean matchesSafely(TrackPayload payload) {
+                    return payload.event().equals("Application Opened")
+                        && payload.properties().getBoolean("from_background", false);
```
Comment (carlos-kelly) [inline analytics/src/test/java/com/segment/analytics/AnalyticsTest.java]: This is the third integration test that verifies when we send the app to the background, then resume it to foreground the `from_background` param is `true`.

File: analytics/src/main/java/com/segment/analytics/AnalyticsActivityLifecycleCallbacks.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+import android.app.Activity;
+import android.app.Application;
+import android.content.pm.PackageInfo;
+import android.os.Bundle;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class AnalyticsActivityLifecycleCallbacks implements Application.ActivityLifecycleCallbacks {
```
Comment (f2prateek) [inline analytics/src/main/java/com/segment/analytics/AnalyticsActivityLifecycleCallbacks.java]: This should be a private class.

Comment (f2prateek) [reply analytics/src/main/java/com/segment/analytics/AnalyticsActivityLifecycleCallbacks.java]: (or at least package protected)

File: analytics/src/test/java/com/segment/analytics/AnalyticsTest.java (hunk 5 above)
Comment (f2prateek) [reply analytics/src/test/java/com/segment/analytics/AnalyticsTest.java]: Got it - shouldn't this test be moved somewhere else then?

Comment (f2prateek) [reply analytics/src/test/java/com/segment/analytics/AnalyticsTest.java]: I see it is moved, never mind!

**Comment by fathyb** [comment]
I merged by accident, sorry! Made a revert PR under #621
